diff --git a/lib/bes/unix/brew/brew_cli_args.py b/lib/bes/unix/brew/brew_cli_args.py
index 6a19cd70..141f4241 100644
--- a/lib/bes/unix/brew/brew_cli_args.py
+++ b/lib/bes/unix/brew/brew_cli_args.py
@@ -26,7 +26,7 @@ class brew_cli_args(object):
     self.__brew_add_common_args(p)
     
     # available
-    p = subparser.add_parser('available', help = 'Print list of available packages.')
+    p = subparser.add_parser('xavailable', help = 'Print list of available packages.')
 
     # files
     p = subparser.add_parser('files', help = 'Print files for a package.')
diff --git a/lib/bes/vmware/vmware_cli_args.py b/lib/bes/vmware/vmware_cli_args.py
index 91cb7b85..b3cc4cfe 100644
--- a/lib/bes/vmware/vmware_cli_args.py
+++ b/lib/bes/vmware/vmware_cli_args.py
@@ -45,7 +45,7 @@ class vmware_cli_args(object):
                    help = 'The script filename [ ]')
 
     # vm_run_program
-    p = subparser.add_parser('vm_run_program', help = 'Run a program in a vm.')
+    p = subparser.add_parser('xvm_run_program', help = 'Run a program in a vm.')
     vmware_options_cli_args.add_arguments(p)
     self.__vmware_add_common_run_program_args(p)
     p.add_argument('vm_id', action = 'store', type = str, default = None,
