diff --git a/lib/bes/btask/btask_process.py b/lib/bes/btask/btask_process.py
index 1102bc7c..71bde873 100755
--- a/lib/bes/btask/btask_process.py
+++ b/lib/bes/btask/btask_process.py
@@ -94,8 +94,13 @@ class btask_process(object):
     
   @classmethod
   def _process_main(clazz, encoded_task_data):
+    import signal
+    signal.signal(signal.SIGINT, signal.SIG_IGN)
+    
     check.check_bytes(encoded_task_data)
 
+
+    
     task_data = pickle.loads(encoded_task_data)
     name = task_data.name
     clazz._log.log_d(f'{name}: task_data={task_data}')
diff --git a/lib/bes/btask/btask_task.py b/lib/bes/btask/btask_task.py
index a1f52e3e..e278a176 100644
--- a/lib/bes/btask/btask_task.py
+++ b/lib/bes/btask/btask_task.py
@@ -16,7 +16,7 @@ class btask_task(namedtuple('btask_task', 'task_id, add_time, config, function,
     config = check.check_btask_config(config)
     check.check_callable(function)
     check.check_dict(args, allow_none = True)
-    check.check_callable(callback)
+    check.check_callable(callback, allow_none = True)
     check.check_callable(progress_callback, allow_none = True)
     
     return clazz.__bases__[0].__new__(clazz,
